Fixes #58

knatten opened the pull request:
Fixes #58

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+§[over.match.best]¶2.5 continues:
+> [or, if not that] F1 and F2 are function template specializations, and the function template for F1 is more specialized than the template for F2 according to the partial ordering rules
+
+I won't go into all the details for partial ordering here, but skip to the interesting parts. In short, we transform each specialization by substituting each template parameter with a unique, made up type and get  `foo₁(X, ...)` and `foo₂(Y)`. Then we do deduction from each transformed function to the other, original template.
```
nit:
```suggestion
I won't go into all the details for partial ordering here, but skip to the interesting parts. In short, we transform each specialization by substituting each template parameter with a unique, made-up type and get `foo₁(X, ...)` and `foo₂(Y)`. Then we do deduction from each transformed function to the other, original template.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+> The types used to determine the ordering depend on the context in which the partial ordering is done:
+> - In the context of a function call, the types used are those function parameter types for which the function call has arguments.
+
+We now do deduction of  `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶10 says:
+
+> During partial ordering, if `Ai` was originally a function parameter pack [and] `Pi` is not a function parameter pack, template argument deduction fails.
```
```suggestion
We now do deduction of  `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶9.2 says:

> During partial ordering, if `A`<sub>*i*</sub> was originally a pack expansion:
>
> — (...) if `P`<sub>*i*<sub> is not a pack expansion, template argument deduction fails.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+
+> During partial ordering, if `Ai` was originally a function parameter pack [and] `Pi` is not a function parameter pack, template argument deduction fails.
+
+`Ai` in this case is `Y`, which was originally a function parameter pack. `Pi` is `T`, which is not a function parameter pack, so deduction fails.
+
+Since we can deduce `foo₂` from `foo₁`, but not `foo₁` from `foo₂`,  `foo₁` is more specialized and a better function. Overload resolution picks it and prints `A`.
```
nit:

```suggestion
`A`<sub>*i*<sub> in this case is `Y`, which was originally a function parameter pack. `P`<sub>*i*<sub> is `T`, which is not a function parameter pack, so deduction fails.

Since we can deduce `foo₂` from `foo₁`, but not `foo₁` from `foo₂`, `foo₁` is more specialized and a better function. Overload resolution picks it and prints `A`.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+
+Now let's look at the call `foo(1,2)`.
+
+Again we start by deducing the template arguments. For `foo₁`, the first argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`.
+
+For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and it's signature is `foo₂(int, int)`.
```
nit:

```suggestion
Again we start by deducing the template arguments. For `foo₁`, the first argument `1` is used to deduce `T` to be `int`. Its signature is then `foo₁(int, ...)`.

For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and its signature is `foo₂(int, int)`.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+
+For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and it's signature is `foo₂(int, int)`.
+
+Which of these is the best match? We turn again to §[over.match.best]¶2 which checks whether
+
+>for some argument *j*, `ICSj(F1)` is a better conversion sequence than `ICSj(F2)`
```
```suggestion
Which of these is the best match? We turn to §[over.match.best.general]¶2.1 which checks whether for some argument of a function call there exists a better implicit conversion sequence from this argument to the corresponding parameter in the other function call.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+
+>for some argument *j*, `ICSj(F1)` is a better conversion sequence than `ICSj(F2)`
+
+For both templates, the first argument is an exact match. For `foo₁`, the second argument `2` requires an ellipsis conversion sequence to match the second parameter `...`: §[over.ics.ellipsis]¶1:
+
+> An ellipsis conversion sequence occurs when an argument in a function call is matched with the ellipsis parameter specification of the function called
```
```suggestion
For both templates, the first argument is an exact match. For `foo₁`, the second argument `2` requires an ellipsis conversion sequence to match the ellipsis parameter specification `...`: §[over.ics.ellipsis]¶1:

> An ellipsis conversion sequence occurs when an argument in a function call is matched with the ellipsis parameter specification of the function called (see [§expr.call]).
```

knatten commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,62 @@
+
+We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`. However, for overload resolution we ignore the `...` since it doesn't have a matching argument: §[over.match.viable]¶2.3:
+
+> For the purposes of overload resolution, the parameter list is truncated on the right, so that there are exactly *m* parameters.
```
Yeah this seems a bit wrong then. I find this pretty hard to wrap my head around. (This question is user-contributed and the explanation was written in collaboration with the contributor.)

I agree that it seems like the ellipsis is not part of the parameter-declaration, but that https://timsong-cpp.github.io/cppwp/std23/dcl.fct#5 says the presence or absence of one is rather part of the parameter-type-list. Then https://timsong-cpp.github.io/cppwp/std23/over.match.viable#2.2 talks about "parameter list", which I don't find defined anywhere. Is the "parameter list" just "everything between the parentheses", including both parameters and a possible ellipsis? So basically the same as "parameter-type-list" but before transformation?

I'd appreciate if you could write up a concrete suggestion here. If not, I could try to email the original author, but it's been five years.

tocic replied:
I've made some suggestions below.

knatten replied:
Thanks! I'm still behind on a lot of stuff here. I squashed the commits now so it's easier to review. All the separate suggestions were hard to work with. Will look into it asap.

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,61 @@
+
+First let's look at the call `foo(1)`. The explanation is lengthy, but rest assured we reuse most of it in the much shorter explanation of `foo(1,2)`! 
+
+We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`. However, for overload resolution we ignore the `...` since it doesn't have a matching argument: §[over.match.viable]¶2.3:
+
+> For the purposes of overload resolution, the parameter list is truncated on the right, so that there are exactly *m* parameters.
+
+So we end up with the viable function `foo₁(int)`.
```
```suggestion
We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. Its signature is then `foo₁(int, ...)`. Is it viable? §[over.match.viable]¶2.1:

> — If there are `m` arguments in the list, all candidate functions having exactly `m` parameters are viable.

But don't we have 2 parameters and 1 argument? It turns out an ellipsis is not considered to be a parameter in this context since it's not included in the *parameter-declaration* grammar production (§[dcl.fct]¶nt:parameter-declaration).

Regarding §[over.match.viable]¶3 and §[over.match.viable]¶4, `foo₁` doesn't have any constraints and there exists an implicit conversion sequence from `1` to `int` (namely, an empty standard conversion sequence since no conversions are required). So `foo₁` is viable.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,61 @@
+
+So we end up with the viable function `foo₁(int)`.
+
+For `foo₂`, `T...` is a function parameter pack. The argument `1`  is used to deduce this as one `int`, and it's signature is `foo₂(int)`.
```
```suggestion
For `foo₂`, `T...` is a function parameter pack. The argument `1` is used to deduce this as one `int`, and its signature is `foo₂(int)`. The same reasoning as above applies, concluding that `foo₂` is viable.
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,61 @@
+>
+> — `F1` and `F2` are function template specializations, and the function template for `F1` is more specialized than the template for `F2` according to the partial ordering rules described in §[temp.func.order] (...)
+
+I won't go into all the details for partial ordering here, but skip to the interesting parts. In short, we transform each specialization by substituting each template parameter with a unique, made-up type and get `foo₁(X, ...)` and `foo₂(Y)`. Then we do deduction from each transformed function to the other, original template.
+
+Again, the ellipsis is ignored since it doesn't have an argument at the call site: §[temp.deduct.partial]¶3:
+
+> The types used to determine the ordering depend on the context in which the partial ordering is done:
+> - In the context of a function call, the types used are those function parameter types for which the function call has arguments.
+
+We now do deduction of  `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶9.2 says:
```
```suggestion
We won't go into all the details for partial ordering here, but skip to the interesting parts. In short, we transform each specialization by substituting each template parameter with a unique, made-up type and get `foo₁(X, ...)` and `foo₂(Y)`.

We now do deduction of `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶9.2 says:
```

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,61 @@
+
+Again we start by deducing the template arguments. For `foo₁`, the first argument `1` is used to deduce `T` to be `int`. Its signature is then `foo₁(int, ...)`.
+
+For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and its signature is `foo₂(int, int)`.
+
```
```suggestion
For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and its signature is `foo₂(int, int)`.

`foo₂` is viable for the same reasons as before, `foo₁` is viable due to §[over.match.viable]¶2.2:

> A candidate function having fewer than `m` parameters is viable only if it has an ellipsis in its parameter list (§[dcl.fct]). For the purposes of overload resolution, any argument for which there is no corresponding parameter is considered to “match the ellipsis” (§[over.ics.ellipsis]).

```

tocic replied:
Here we can also say that for `2` -> `...` we use an ellipsis conversion sequence instead of an empty standard conversion sequence. But perhaps the explanation is already big enough.

knatten replied:
Yeah, this is long enough already imo. These are supposed to be instructive, detailed enough to be complete, but not too off-putting for people to actually read.

knatten commented:
Thanks a lot for all the help with this one! I made two small changes:

I think it's better to say that the ellipsis is not part of the _[parameter-declaration-list](https://timsong-cpp.github.io/cppwp/std23/dcl.fct#nt:parameter-declaration-list)_,  since the _parameter-declaration-clause_ is divided into the _parameter-declaration-list_ and the ellipsis.

I removed the references to the standard that we don't quote since I found those confusing. It made me skim up to see where they had been previously discussed.  And I simplified that part a little bit more.

tocic commented on questions/250/explanation.md:
```diff
@@ -1,62 +1,60 @@
+
+For `foo₂`, `T...` is a function parameter pack. The argument `1` is used to deduce this as one `int`, and its signature is `foo₂(int)`. The same reasoning as above applies, concluding that `foo₂` is viable.
+
+So, both `foo`s are viable. Which one is best? §[over.match.best]¶2:
```
```suggestion
So, both `foo`s are viable. Which one is best? §[over.match.best.general]¶2:
```
